[feature/APPC-1291] - Implement min and max limit and default values chips

joaopmgouveia opened the pull request:
**What does this PR do?**

   This PR implements the changes to the Top Up screen related to the minimum and maximum allowed values and the default values chips, that allow the user to pick a predetermined value.

Design: 
1- zpl.io/bz7GBAA
2- zpl.io/2jWl1Er            

**Database changed?**

   No

**Where should the reviewer start?**

- TopUpFragment.kt
- TopUpFragmentPresenter.kt
- TopUpValuesService.kt

**How should this be manually tested?**

  Flow on how to test this or QA Tickets related to this use-case: [APPC-1291](https://aptoide.atlassian.net/browse/APPC-1291)

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-1291](https://aptoide.atlassian.net/browse/APPC-1291)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new libs, new keys, etc.)




**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

joaopmgouveia committed:
[feature/APPC-1291] - 1- Create MockedTopUpValuesService to mock service responses with values needed for the view; 2- Add invalid warning icon and TextView in "fragment_top_up" layout; 3- Create "handleShowValueWarning", "fetchMinAndMaxValues" and "showValueWarning" methods in TopUpFragmentPresenter so that with every change in the input EditText the presenter can check if it is a valid value; 4- Create "showMaxValueWarning", "showMinValueWarning" and "hideValueInputWarning" methods so the presenter is able to display/hide the correct warning

joaopmgouveia committed:
[feature/APPC-1291] - 1- Add "changeMainValueColor" method in TopUpFragment so that the presenter can change the color of the main value text depending on the input value; 2- Add logic in TopUpFragmentPresenter so that the "NEXT" button and color of the input text can be enabled/disabled and change colors respectively

joaopmgouveia committed:
Merge branch 'dev' of https://github.com/Aptoide/appcoins-wallet-android into feature/APPC-1291-show_min_max_top_up_value

joaopmgouveia committed:
Merge branch 'dev' of https://github.com/Aptoide/appcoins-wallet-android into feature/APPC-1291-show_min_max_top_up_value

joaopmgouveia committed:
feature/APPC-1291[] - 1- Create "chip_selected_background" and "chip_unselected_background" drawables to match the design for the chips; 2- Add "default_chips_layout" in "fragment_top_up" layout to add the chips to the Top Up screen

joaopmgouveia committed:
Merge branch 'dev' of https://github.com/Aptoide/appcoins-wallet-android into feature/APPC-1291-show_min_max_top_up_value

joaopmgouveia committed:
[feature/APPC-1291] - 1- Create "chip_selected_disabled_background" and "chip_unselected_disabled_background" drawable so that we can change the background of disabled chips; 2- Add logic in TopUpFragment so TopUpFragmentPresenter can handle the chip's input and the initial setup; 3- Create "setChipOnClickListener" method so that each click emits a different value for different chips clicked; 4- Change "navigateToPayment" method in TopUpActivity and "newInstance" method in PaymentAuthFragment so that when clicking on the "NEXT" button the values and the selected chip can be passed to this fragment

joaopmgouveia committed:
[feature/APPC-1291] - 1- Create "handlePreselectedChip" method in TopUpFragmentPresenter so that a chip is preselected beforehand

joaopmgouveia committed:
[feature/APPC-1291] - 1- Change "hideKeyboard" and create "paymentMethodsFocusRequest" methods in order to hide the keyboard when entering the TopUp screen and when selecting the payment method; 2- Change "changeMainValueText" so that when clicking a chip the cursor is always pushed to the end of the value, so that a user can erase the input value without having to change the cursor's position

joaopmgouveia committed:
[feature/APPC-1291] - 1- Create "disableSwapCurrency" and "enableSwapCurrency" methods in TopUpFragmentView so that the presenter can disable and enable the swap currency button to avoid misinputs; 2- Create "chipValuesList" lateinit variable in TopUpFragmentPresenter to avoid multiple requests; 3- Create "handleManualInputValue" method in TopUpFragmentPresenter so that the presenter can select/unselect chips based on input written by the user

joaopmgouveia committed:
[feature/APPC-1291] - 1- Delete MockedTopUpValuesService; 2- Create "TopUpValuesService", "TopUpDefaultValuesResponse", "TopUpLimitValue", "TopUpLimitValuesResponse" and "TopUpValuesApiResponseMapper" to request and map backend information related to TopUp limit and default values

joaopmgouveia committed:
[feature/APPC-1291] - Add logic on method "onViewCreated" in TopUpFragment, and on method "setupUi" in TopUpFragmentPresent to cope with the view's initial setup and posterior device rotations

joaopmgouveia committed:
[feature/APPC-1291] - 1- Change parameter name on "initialInputSetup" method in TopUpFragmentView; 2- Create NumberFormatterUtils to cope with some currencies having default values chips with values too big for the chip

joaopmgouveia committed:
[feature/APPC-1291] - 1- Change "setChipChecked" method to the correct "selectChip" method in TopUpFragment; 2- Move "handleManualInputValue" method closer to the start of the chain in TopUpFragmentPresenter so that when selecting a different chip no flicker happens between both chips clicked

joaopmgouveia committed:
Merge branch 'dev' of https://github.com/Aptoide/appcoins-wallet-android into feature/APPC-1291-show_min_max_top_up_value

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/di/ToolsModule.java:
```diff
@@ -1015,8 +1017,30 @@ AsfInAppPurchaseInteractor provideAsfInAppPurchaseInteractor(
+        .create(TopUpValuesService.TopUpValuesApi.class);
+  }
+
+  @Provides TopUpValuesApiResponseMapper providesTopUpValuesApiResponseMapper() {
```
@Singleton

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpActivity.kt:
```diff
@@ -86,15 +87,18 @@ class TopUpActivity : BaseActivity(), TopUpActivityView, ToolbarManager, UriNavi
-                                 transactionType: String, bonusValue: String) {
+                                 transactionType: String, bonusValue: String,
+                                 selectedChip: Int, chipValue1: FiatValue, chipValue2: FiatValue,
+                                 chipValue3: FiatValue, chipValue4: FiatValue) {
```
I know what I said at the time :/, but I searched and I found a way to pass an array in a bundle, check the 2nd answer on the link that uses Serializable.
https://stackoverflow.com/questions/12092612/pass-list-of-objects-from-one-activity-to-other-activity-in-android

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -158,4 +186,116 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
+  private fun getChipValuesList(): List<FiatValue> {
+    return interactor.getDefaultValues()
+        .subscribeOn(networkScheduler)
+        .blockingGet()
```
Why use blockingGet() here?

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -21,29 +21,43 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
 
   private val disposables: CompositeDisposable = CompositeDisposable()
 
+  private lateinit var chipValuesList: List<FiatValue>
```
Correct me if I’m wrong. You did this to cache the values as they will never change (at least not on short term). If that’s what you’re doing, you should do it on the repo (but as there’s no specific repo for topup and there’s already a bdsrepo that is the repo being used, I would move this logic to the Interactor).
Then every time the presenter needs the list, the presenter requests the list to the interactor and the interactor how it gives the list

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -158,4 +186,116 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
+    if (inputValue != "--") {
+      val value = inputValue.toBigDecimal()
+      for (i in chipValuesList.indices) {
+        if (value == chipValuesList[i].amount) {
```
Mainly do to this method I would change this from a list to map, that would map the values to the index, this way you avoid to cycle through the list every time you need the index

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragment.kt:
```diff
@@ -107,15 +116,23 @@ class TopUpFragment : DaggerFragment(), TopUpFragmentView {
     topUpActivityView?.showToolbar()
-    presenter.present(appPackage)
-
+    if (savedInstanceState?.containsKey(SELECTED_CHIP) == true) {
```
this can go to the lateinit variables init (like appPackage). What you need here to use the boolean is the savedInstance != null or not. Worst case you can check if the variable that would be created is != -1 or not. Just to remove variable instantiation from here

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragment.kt:
```diff
@@ -264,6 +293,184 @@ class TopUpFragment : DaggerFragment(), TopUpFragmentView {
+
+  override fun changeMainValueColor(isValid: Boolean) {
+    if (isValid) {
+      main_value.setTextColor(ContextCompat.getColor(context!!, R.color.black))
```
context?.let{ if(....) ... }

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragment.kt:
```diff
@@ -264,6 +293,184 @@ class TopUpFragment : DaggerFragment(), TopUpFragmentView {
+  }
+
+  override fun getSelectedChip(): Int {
+    var selectedChip: Int = -1
```
No need to declare it as Int

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragment.kt:
```diff
@@ -264,6 +293,184 @@ class TopUpFragment : DaggerFragment(), TopUpFragmentView {
+    default_chip1.isChecked = false
+    default_chip2.isChecked = false
+    default_chip3.isChecked = false
+    default_chip4.isChecked = false
```
Since you usually use all 4 chips at once (assigning them with the same values), you could create a list with them onViewCreated

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -63,15 +77,20 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
               it.currency.appcValue != DEFAULT_VALUE && it.currency.fiatValue != DEFAULT_VALUE
             }.doOnNext {
               view.showLoading()
+              val values = getChipValuesList()
```
Use this with a chain, no need to block and get

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -158,4 +186,116 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
+  private fun handleInvalidValueInput(packageName: String, topUpData: TopUpData) {
+    if (topUpData.currency.fiatValue != DEFAULT_VALUE) {
+      val value =
+          FiatValue(BigDecimal(topUpData.currency.fiatValue), topUpData.currency.fiatCurrencyCode,
```
I would be careful with this as this was originally a BigDecimal turned to String and then it’s being converted to BigDecimal again. This already gave us problems once, due to BigDecimal not handling “,” very well

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -158,4 +186,116 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
+
+  private fun fetchMinAndMaxValues(): Single<Pair<FiatValue, FiatValue>> {
+    return Single.zip(interactor.getMaxTopUpValue(), interactor.getMinTopUpValue(),
+        BiFunction { maxValue: FiatValue, minValue: FiatValue -> Pair(maxValue, minValue) })
```
I would pass the zip logic to the interactor and have the presenter call a method that would return an object with both values. (Keep both methods as private, as they can later be used as public if we only need one of the values)

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -158,4 +186,116 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
+
+  private fun handleValueChipsClick() {
+    disposables.add(view.getChipsClick()
+        .doOnNext { index ->
```
Add a observeOn, just to absolutely guarantee that this will not run on the io thread (as it is right now it should work, but there’s a small probability that it can’t)

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -158,4 +186,116 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
+
+  private fun handleChipFiatInput(value: FiatValue) {
+    view.changeMainValueText(value.amount.toString())
+  }
```
No need for a 1 line method

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpLimitValue.kt:
```diff
@@ -0,0 +1,5 @@
+
+import com.asfoundation.wallet.ui.iab.FiatValue
+
+data class TopUpLimitValue(val value: FiatValue)
```
This is not needed you can map it to FiatValue

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/payment/PaymentAuthFragment.kt:
```diff
@@ -132,18 +174,31 @@ class PaymentAuthFragment : DaggerFragment(), PaymentAuthView {
-                    bonusValue: String): PaymentAuthFragment {
+                    bonusValue: String, selectedChip: Int,
+                    chipValue1: FiatValue, chipValue2: FiatValue, chipValue3: FiatValue,
+                    chipValue4: FiatValue): PaymentAuthFragment {
```
Same as above, try to pass it as a list

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/payment/PaymentAuthFragment.kt:
```diff
@@ -423,4 +491,58 @@ class PaymentAuthFragment : DaggerFragment(), PaymentAuthView {
+    default_chip1.text = chipValue1.symbol + chipValue1.amount
+    default_chip2.text = chipValue2.symbol + chipValue2.amount
+    default_chip3.text = chipValue3.symbol + chipValue3.amount
+    default_chip4.text = chipValue4.symbol + chipValue4.amount
```
Same as before, try to do a list with the views

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/payment/PaymentAuthFragment.kt:
```diff
@@ -423,4 +491,58 @@ class PaymentAuthFragment : DaggerFragment(), PaymentAuthView {
+
+  private fun setSelectedChipText(index: Int) {
+    when (index) {
+      0 -> default_chip1.setTextColor(ContextCompat.getColor(context!!, R.color.white))
```
context?.let {...} do the same for the TopUpFragment

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/util/NumberFormatterUtils.kt:
```diff
@@ -0,0 +1,29 @@
+    val divideBy = fetchLowestValueSuffix.key
+    val suffix = fetchLowestValueSuffix.value
+
+    val truncatedValue = value / (divideBy!! / 10)
```
This can’t be null

joaopmgouveia committed:
[feature/APPC-1291] - 1- Change endpoints in TopUpValuesService; 2- Create "formatDecimalPlaces" and "removeDecimalPlaces" methods and change "formatNumberWithSuffix" method so that numbers with decimal places with zeroes are removed; 3- Change from Int to String in Fiat data class parameter in TopUpDefaultValueBody to handle values with decimal places

joaopmgouveia committed:
[feature/APPC-1291] - 1- Change individual chip values passed to PaymentAuthFragment to a chip value list; 2- Add "singleton" tag to "providesTopUpValuesService" method in ToolsModule

joaopmgouveia committed:
[feature/APPC-1291] - Delete unnecessary data class TopUpLimitValue

joaopmgouveia committed:
[feature/APPC-1291] - 1- Create "chipViewList" in PaymentAuthFragment and TopUpFragment to better organize the code when interacting with the chips

joaopmgouveia committed:
[feature/APPC-1291] - 1- Remove unnecessary special symbol "!!" in NumberFormatterUtils; 2- Remove unused variable "NUMBER_OF_CHIPS" in TopUpFragment
; 3- Remove unnecessary Int cast in TopUpFragment; 4- Remove "handleChipFiatInput" method in TopUpFragmentPresenter

joaopmgouveia committed:
[feature/APPC-1291] - 1- Create TopUpLimitValues data class to substitute the FiatValue Pair used in the "fetchMinAndMaxValues" method in TopUpFragmentPresenter, now after the mapping both values are mapped into the same object and accessed from it

joaopmgouveia committed:
[feature/APPC-1291] - 1- Change cache logic from TopUpFragmentPresenter to TopUpInteractor, by removing "chipValuesList" from the presenter and by adding "chipValueIndexMap" to the interactor parameters; 2- Move "getChipIndex" and "getChipValuesList" logic to interactor; 3- Remove "getChipValuesList" method's blockingGet

joaopmgouveia committed:
Merge branch 'dev' of https://github.com/Aptoide/appcoins-wallet-android into feature/APPC-1291-show_min_max_top_up_value

joaopmgouveia committed:
[feature/APPC-1291] - Fix issue where the keyboard would appear and disappear when entering the Top Up screen

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -117,15 +125,30 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
-      view.selectChip(chipIndex)
-    } else {
-      view.unselectChips()
+    if (topUpData.currency.fiatValue != "--") {
```
Extract this If to outside of the function

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -158,4 +209,92 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
   }
+
+  private fun handleInvalidValueInput(packageName: String, topUpData: TopUpData) {
+    if (topUpData.currency.fiatValue != DEFAULT_VALUE) {
```
Extract if to outside of the function

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt:
```diff
@@ -275,27 +292,9 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
-  private fun getChipValuesList(): List<FiatValue> {
+  private fun getChipValue(index: Int): Single<FiatValue> {
     return interactor.getDefaultValues()
         .subscribeOn(networkScheduler)
```
This subscribe logic should be on the chain that calls this and not here (it already is anyways)

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpInteractor.kt:
```diff
@@ -51,18 +53,36 @@ class TopUpInteractor(private val repository: BdsRepository,
-        .map { it.value }
+  fun getLimitTopUpValue(): Single<TopUpLimitValues> {
+    return topUpValuesService.getLimitValues()
+        .map { it }
```
This is not required

JDSM01 commented on app/src/main/java/com/asfoundation/wallet/topup/TopUpInteractor.kt:
```diff
@@ -51,18 +53,36 @@ class TopUpInteractor(private val repository: BdsRepository,
+      Single.just(ArrayList(chipValueIndexMap.keys))
+    } else {
+      topUpValuesService.getDefaultValues()
+          .map { it }
```
Not required

joaopmgouveia committed:
[feature/APPC-1291] - 1- Extract condition to outside of the function on "handleManualInputValue" and "handleInvalidValueInput" methods in TopUpFragmentPresenter; 2- Remove unnecessary "subscribeOn" on "getChipValue" method in TopUpFragmentPresenter; 3- Remove unnecessary "map" functions on "getLimitTopUpValue" and "getDefaultValues" methods in TopUpInteractor; 4- Rearrange code on "handlePaymentMethodSelected" method in TopUpFragmentPresenter

joaopmgouveia committed:
[feature/APPC-1291] - 1- Change "map" fuctions to "doOnSuccess" and "doOnNext" in certain methods in TopUpFragmentPresenter; 2- Add "throttleFirst" on "handleValueChipsClick" method in TopUpFragmentPresenter, so that flickers when clciking more than 1 chip don't occur

joaopmgouveia committed:
[feature/APPC-1291] - 1- Add logic to "handleManualAmountChange" method so that when the selected currency is APPC-C and when the main value changes all chips are unselected

joaopmgouveia committed:
Merge branch 'dev' of https://github.com/Aptoide/appcoins-wallet-android into feature/APPC-1291-show_min_max_top_up_value

joaopmgouveia committed:
[feature/APPC-1291] - 1- Create "setUnselectedChipsBackground" method in TopUpFragment so that the presenter can change the background and the text colour without unchecking the chip when changing to APPC-C currency; 2- Modify "handleChangeCurrencyClick" method in TopUpFragmentPresenter so that when changing between currencies the presenter can select/unselect chip accordingly

joaopmgouveia committed:
[feature/APPC-1291] - Add logic to "handleNextClick" in TopUpFragmentPresenter so that when clicking the next button with APPC-C as selected currencyall chips are disabled in the payment details screen
